storage/src/backend/connection.rs

```diff
@@ -428,90 +450,110 @@ impl Connection {
+
+                for (key, value) in mirror.config.headers.iter() {
                     headers.insert(
                         HeaderName::from_str(key).unwrap(),
                         HeaderValue::from_str(value).unwrap(),
                     );
                 }
-            }
 
-            let result = self.call_inner(
-                &self.client,
-                method.clone(),
-                current_url.to_string().as_str(),
-                &query,
-                data_cloned,
-                headers,
-                catch_status,
-                false,
-            );
-
-            match result {
-                Ok(resp) => {
-                    if resp.status() < StatusCode::INTERNAL_SERVER_ERROR {
-                        return Ok(resp);
-                    }
-                }
-                Err(err) => {
-                    warn!(
-                        "request mirror server failed, mirror: {},  error: {:?}",
-                        format!("{:?}", mirror).to_lowercase(),
-                        err
-                    );
-                    mirror.failed_times.fetch_add(1, Ordering::Relaxed);
+                let current_url = mirror.mirror_url(url)?;
+                debug!("mirror server url {}", current_url);
 
-                    if mirror.failed_times.load(Ordering::Relaxed) >= mirror.failure_limit {
-                        warn!(
-                            "reach to fail limit {}, disable mirror: {}",
-                            mirror.failure_limit,
-                            format!("{:?}", mirror).to_lowercase()
-                        );
-                        mirror.status.store(false, Ordering::Relaxed);
+                let result = self.call_inner(
+                    &self.client,
+                    method.clone(),
+                    current_url.as_str(),
+                    &query,
+                    data_cloned,
+                    headers,
+                    catch_status,
+                    false,
+                );
 
-                        let mut idx = 0;
-                        loop {
-                            if idx == self.mirror_state.mirrors.len() {
-                                break None;
+                match result {
+                    Ok(resp) => {
+                        // If the response status >= INTERNAL_SERVER_ERROR, fall back to original registry.
+                        if resp.status() < StatusCode::INTERNAL_SERVER_ERROR {
+                            // If the response status is not UNAUTHORIZED, we do not need to get the token URL
+                            // from "www-authenticate" in response headers.
+                            if resp.status() != StatusCode::UNAUTHORIZED {
+                                return Ok(resp);
+                            }
+
+                            // Remove HEADER_AUTHORIZATION to get correct response with "www-authenticate"
```

Comment:
Why do we need to handle the expired auth header in `connection.rs` ? It has been handled in `registry.rs`.

Reply:
You are right, I am moving it to the `registry.rs`.

Reply:
Done.